crates/storage/src/counter.rs

```diff
@@ -1,57 +1,98 @@
 use {
-    crate::{Borsh, Codec, Item},
-    grug_types::{Number, NumberConst, StdResult, Storage},
+    crate::{Borsh, Codec, Item, Key, Map},
+    grug_types::{Number, StdResult, Storage},
 };
 
-/// An abstraction over `Item`. Stores a single number that is monotonically
-/// incremented one unit at a time.
+/// A single number that is monotonically incremented by the given step size.
+///
+/// In internally, this is an abstraction over an [`Item`](crate::Item).
 pub struct Counter<'a, T, C = Borsh>
 where
     C: Codec<T>,
 {
     item: Item<'a, T, C>,
+    base: T,
+    step: T,
 }
 
 impl<'a, T, C> Counter<'a, T, C>
 where
+    T: Number + Copy,
     C: Codec<T>,
 {
-    pub const fn new(storage_key: &'a str) -> Self {
+    pub const fn new(storage_key: &'a str, base: T, step: T) -> Self {
         Self {
             item: Item::new(storage_key),
+            base,
+            step,
         }
     }
 
     /// Load the current counter value.
-    ///
-    /// Error if the counter has not been initialized.
-    pub fn load(&self, storage: &dyn Storage) -> StdResult<T> {
-        self.item.load(storage)
+    pub fn current(&self, storage: &dyn Storage) -> StdResult<T> {
+        self.item
+            .may_load(storage)
+            .map(|maybe_value| maybe_value.unwrap_or(self.base))
+    }
+
+    /// Increment the value by the step size; return the values before and after
+    /// incrementing.
+    pub fn increment(&self, storage: &mut dyn Storage) -> StdResult<(T, T)> {
+        let old_value = self.current(storage)?;
+        let new_value = old_value.checked_add(self.step)?;
+
+        self.item.save(storage, &new_value)?;
+
+        Ok((old_value, new_value))
     }
 }
 
-impl<'a, T, C> Counter<'a, T, C>
+/// A single number under each key, that is monotonically incremented by the
+/// given step size.
+///
+/// Internally, this is an abstraction over a [`Map`](crate::Map).
+pub struct Counters<'a, K, T, C = Borsh>
 where
-    T: Number + NumberConst,
     C: Codec<T>,
 {
-    /// Initialize the incrementor to the zero value.
-    ///
-    /// This is typically done during contract instantiation.
-    pub fn initialize(&self, storage: &mut dyn Storage) -> StdResult<()> {
-        self.item.save(storage, &T::ZERO)
+    map: Map<'a, K, T, C>,
+    base: T,
+    step: T,
+}
+
+impl<'a, K, T, C> Counters<'a, K, T, C>
+where
+    C: Codec<T>,
+{
+    pub const fn new(storage_key: &'a str, base: T, step: T) -> Self {
+        Self {
+            map: Map::new(storage_key),
+            base,
+            step,
+        }
+    }
+}
+impl<'a, K, T, C> Counters<'a, K, T, C>
+where
+    K: Key + Copy,
```

Comment:
`K` needs to be `Copy`, but we typically use reference `&T` as keys, which is indeed `Copy`-able.